src/Rules/PHPUnit/ClassNamingRule.php

```diff
@@ -0,0 +1,111 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\PHPUnit;
+
+use PhpParser\Node;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ReflectionProvider;
+use PHPStan\Rules\IdentifierRuleError;
+use PHPStan\Rules\Rule;
+use PHPStan\Rules\RuleErrorBuilder;
+use PHPUnit\Framework\TestCase;
+use function sprintf;
+use function strlen;
+use function substr_compare;
+
+/**
+ * @implements Rule<Node\Stmt\Class_>
+ */
+class ClassNamingRule implements Rule
+{
+
+	private ReflectionProvider $reflectionProvider;
+
+	public function __construct(ReflectionProvider $reflectionProvider)
+	{
+		$this->reflectionProvider = $reflectionProvider;
+	}
+
+	public function getNodeType(): string
+	{
+		return Node\Stmt\Class_::class;
+	}
+
+	public function processNode(Node $node, Scope $scope): array
+	{
+		if (!isset($node->namespacedName)) {
+			return [];
+		}
+
+		$class = $this->reflectionProvider->getClass($node->namespacedName->toString());
+
+		if (!$class->isSubclassOf(TestCase::class)) {
+			return [];
+		}
+
+		$errors = [];
+
+		if ($class->isAbstract()) {
+			$this->requireSuffix(
+				$errors,
+				$class->getName(),
+				'TestCase',
+				'Abstract test case class, \'%s\', should be named ending in \'%s\'.',
+			);
+
+			return $errors;
+		}
+
+		$this->requireSuffix(
+			$errors,
+			$class->getName(),
+			'Test',
+			'Concrete test class, \'%s\', should be named ending in \'%s\'.',
+		);
+
+		if (!$class->isFinal()) {
+			$errors[] = RuleErrorBuilder::message(sprintf(
+				'Concrete test class, \'%s\', should be declared final.',
+				$class->getName(),
+			))->identifier('phpunit.naming')->build();
+		}
```

Comment:
I am not sure this "isFinal" thing is something which should be mixed with a "general naming rule".

whether tests should be final or not is a pretty opinionated decision IMO